### DISCUSSION TITLE
How to use @mswjs/interceptors correctly

### DISCUSSION THREAD

**Discussion by AndonMitev** [vercel/next.js#57749] [Help]
### Summary

Hello, trying to catch all fetch requests that have been made with @mswjs/interceptors.

This is my implementation:

```
import { FetchInterceptor } from '@mswjs/interceptors/fetch';

const interceptor = new FetchInterceptor();

// const interceptor = new FetchInterceptor();

interceptor.apply();

interceptor.on('request', ({ request, requestId }) => {
  console.log(request.method, request.url);
});

interceptor.on(
  'response',
  ({ response, isMockedResponse, request, requestId }) => {
    console.log(
      'response to %s %s was:',
      request.method,
      request.url,
      response
    );
  }
);

```

Problem is that when i send request is like logging even pages, this is how 1 log looks like:

<img width="1017" alt="image" src="https://github.com/vercel/next.js/assets/32274987/3e3bdc11-8569-46fc-b4ec-2c0828a83382">

Called 8 times and im interested only in the final request

<img width="772" alt="image" src="https://github.com/vercel/next.js/assets/32274987/17285e36-63d9-4188-a6bb-09d3fb73a4e6">

Does this mean we need to filter all requests and listen for these that request.url === api_url?

Also if i try to add some of the presets like this one:

`import nodeInterceptors from '@mswjs/interceptors/presets/node'` or this one:
`import browserInterceptors from '@mswjs/interceptors/presets/browser'`

says both of them are not found in the package hmm. Any ideas how to improve the usage to catch only external calls and skip pages icons etc etc

PS: Yes manage to limit it to BE calls only with this

```
if (request.url.startsWith(env.NEXT_PUBLIC_APP_URL)) {
    console.log(request.method, request.url);
  }
```

but this is tighted to only 1 backend but i guess if we want we can extend it manually, finally i think i'm seeing the light in the end of the tunnel :P.

Ok just 1 final thing, how to filter responses that are ssr. I mean i'm doing get to `https://api.themoviedb.org/3/trending/movie/day` and it's fine i can filter that in request interceptor, but then when i receive the response is not coming from the same domain, rather is coming from `http://localhost:3000/movies?_rsc=4rewo`, any idea how to intercept the

**Comment by icyJoseph** [comment]
- Does this mean we need to filter all requests and listen for these that request.url === api_url?

Yes, as they say in their docs:

> [What this library doesn't do](https://www.npmjs.com/package/@mswjs/interceptors#what-this-library-doesnt-do)
> Does not provide any request matching logic;
> Does not decide how to handle requests.

If you did monkey patch yourself, you'd most likely need to also filter out requests.

Could you check that it works in Firefox?

**Reply by AndonMitev** [reply]
Checking bro, meanwhile i think i could handle it like this:

```
import { FetchInterceptor } from '@mswjs/interceptors/fetch';
import { env } from '..';

const interceptor = new FetchInterceptor();

interceptor.apply();

interceptor.on('request', ({ request, requestId }) => {
  if (
    request.url.startsWith(env.NEXT_PUBLIC_APP_URL) ||
    request.url.startsWith('https://api.themoviedb.org/3')
  ) {
    console.log(request.method, request.url);
  }
});

interceptor.on(
  'response',
  ({ response, isMockedResponse, request, requestId }) => {
    console.log('what', response.url);
    if (
      request.url.startsWith(env.NEXT_PUBLIC_APP_URL) ||
      request.url.startsWith('https://api.themoviedb.org/3')
    ) {
      console.log(request.url, response.status);
    }
  }
);

```

For the responses that are coming as server components, response.url is not pointing to the resource rather is pointing to `http://localhost:3000/movies?_rsc=4rewo`, so i think this workaround with checking request url in response interceptor will do the trick

**Reply by AndonMitev** [reply]
@icyJoseph, yeah i can confirm that logs are visible in firefox as well, attaching photo:

<img width="1058" alt="image" src="https://github.com/vercel/next.js/assets/32274987/85bb21b7-de58-47ff-9640-fcef0f2b2db0">

Using firefox: 119.0

**Reply by AndonMitev** [reply]
btw any idea how to mutate headers:

<img width="906" alt="image" src="https://github.com/vercel/next.js/assets/32274987/870bf1b6-414a-4b61-825c-605e7cdf54c8">

I want to pick access token from next-auth session, but not allowing me to set `Authorization: Bearer ${accessToken}`, says headers are readonly :(.

PS: nvm i found it, it uses Set under the hood:

```
request.headers.append('Authorization', `Bearer: ${session?.user.accessToken}`)
```

**Reply by AndonMitev** [reply]
Actually has some weird behavior, when reading `const session = await getSession();` stack exceeds max limit, how come lol

**Reply by icyJoseph** [reply]
Could you clone the request? or make a new one of the original one, but with some/none/modified headers?

**Reply by AndonMitev** [reply]
Ahh i understood the problem, thats tricky one i think

```
// const session = useSession();
  const session = await getSession();
```

Some requests are client, like register user, update password and so on, but other like fetch movies are server component requests and this one is messing with the session, any ideas how to select by the current env, or there is another interceptor that i need to use hmm
